[filebeat][httpjson] Extend filebeat httpjson input for chained calls

kush-elastic opened the pull request:
- Enhancement
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

- This PR contains code for extending httpjson to run chain calls and collect information from multiple urls.
- This httpjson code enables user perform chain in single manner.
<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

## Why is it important?

- Current filebeat httpjson input doesn't support chain calls.
For Example:
       Salesforce: To collect eventLogFile from salesforce, we first need to collect logfile ids from Salesforce REST API and then use those logfile ids to collect more information from Salesforce.

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [X] Check with Salesforce REST API to collect EventLogFile.



<!-- Recommended
## How to test this PR locally
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues
- Relates #22590
<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Requires #123
- Superseds #123
-->

## Use cases
- Configure chain parameters with each chain in step parameter.
<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

kush-elastic committed:
[filebeat][httpjson] Extend filebeat httpjson input for chained calls

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,19 +184,102 @@ func newRequester(
 
-	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, httpResp)
+// generateNewUrl returns new url value using replacement from oldUrl with str ids
+func generateNewUrl(replacement string, oldUrl string, str []byte) (url.URL, error) {
```
This function returns URL and error. Is there an option to return *URL and error? In case of failure, it could return a nil.

kush-elastic replied:
I am not sure but I will look into it.
Thanks

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,19 +184,102 @@ func newRequester(
+	// perform each requestFactory
+	for i, rf := range r.requestFactory {
+		// iterate over collected ids from last response
+		if i != 0 {
```
Why do you need to skip the 0's element in the array?

kush-elastic replied:
0th element will be a regular call you can find in else part. 
0 - regular call
1 - chain step-1
2 - chain step-2
3 and so on...

mtojek replied:
A comment please :)

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,19 +184,102 @@ func newRequester(
+func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher) error {
+	var (
+		httpResp       *http.Response
+		ArrayhttpResp  []*http.Response
```
Is there an option to have only the array? For single elements, it would an array with one element.

kush-elastic replied:
Oh Sorry, it will not work as I am performing some other tasks from single elements so it will be difficult to locate only the required resp.

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,19 +184,102 @@ func newRequester(
+
+	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, finalHttpResps)
 	if err != nil {
 		return err
```
errors.Wrap?

kush-elastic replied:
I am not sure but maybe it's intentional as it was there from before I updated the main method, but I will add some context there.

andrewkroh committed:
Add info to aws-s3 log messages (#29328)

This adds elapsed_time_ns into S3 object read errors and SQS processing warning log messages. This gives context about how long it took to fail. Extended times could be an indicator of a blocked output.

It also adds the number of S3 records contained in the SQS notification to error messages. This gives context about how many S3 objects were being processed when a failure occurs.

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -207,7 +311,99 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
 }
+
+// getJSON returns array of string values from json string
+func getJSON(b, str string) ([]string, error) {
```
Could you please extract it to a helper or any other place?

kush-elastic replied:
Yeah, Right.
I will do it. Thanks!

mtojek commented on x-pack/filebeat/input/httpjson/response.go:
```diff
@@ -74,63 +74,68 @@ func newResponseProcessor(config *responseConfig, pagination *pagination, log *l
 
-func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *transformContext, resp *http.Response) (<-chan maybeMsg, error) {
+func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *transformContext, resp []*http.Response) (<-chan maybeMsg, error) {
 	ch := make(chan maybeMsg)
```
As there is a channel involved, could you please cover it with positive and negative tests?

kush-elastic replied:
Sure, I will work on it.
Thanks

mtojek reviewed (commented):
First review round done.

mtojek commented:
Could you please also invite Marc for reviewing these changes?

ChrsMark committed:
Fix k8s metadata regression (#29329)

kvch committed:
Overwrite index name in index template correctly (#29299)

andrewkroh committed:
Fix decode_cef handling of escaped LF and CR (#29268)

* Fix decode_cef handling of escaped LF and CR

Escaped LF and CR were not handled inside of CEF extension value.
Those characters are now accepted and replaced with their unescaped
equivalents.

Replace fgoto with fnext to fix generated Go code that had duplicate 'goto' statements.

Fixes #16995

andrewkroh committed:
Update cloud.google.com/go/pubsub and supporting libraries (#29353)

Moving from v1.3.1 to v1.17.1.

Updating via `go get cloud.google.com/go/pubsub`.

Relates #29352

ChrsMark committed:
Update k8s autodiscover metadata in nodes earlier (#29289)

narph committed:
Agent configuration overridden by default fleet config (#29297)

* replace config

* changelog

* add test on merge

* fmt

fearful-symmetry committed:
Remote global hostfs variables (#29138)

* first pass at fixing hostfs

* forgot to commit a bunch of stuff, oops

* fix system/linux tests

* fix tests

* fix other tests

* fix diskio tests

* take another pass at fixing diskio tests

* revert to non-breaking change

* fix docs

* fix metricbeat legacy config logic, fix a few tests

* fix import

* fix libraries, docs

* fix processors

* few code changes

Commit:
[automation] update elastic stack version for testing 8.1.0-30a394f1 (#29275)

* [Automation] Update elastic stack version to 8.1.0-30a394f1 for testing

* Reconfigure compilation rate

* Update golden files

* Skip failing test

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>
Co-authored-by: Jaime Soriano Pastor <jaime.soriano@elastic.co>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

Commit:
[Automation] Update elastic stack version to 8.1.0-b6e7e1f3 for testing (#29387)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

mergify[bot] committed:
fix: check the activation script exists on Python virtual environments (#29265) (#29288)

(cherry picked from commit 52e87f02c416fd5d13ab45ad74848fe70b310639)

Co-authored-by: Ivan Fernandez Calvo <kuisathaverat@users.noreply.github.com>

andresrc committed:
Forward port 7.16.0 changelog to master (#29391)

joereith committed:
Metricbeat fails to build linux/mips* targets on linux/amd64 hosts (#28323)

* Update io_helper_linux.go

This edit allows for cross compiling linux/mips* targets on linux/amd64 hosts. before this edit I was getting the following error with the following command:
```bash
user@user:~/Documents/projects/beats/metricbeat$ GOOS=linux GOARCH=mipsle GOMIPS=softfloat go build -trimpath -ldflags="-s -w" -o metricbeat_mipsle
# github.com/elastic/beats/v7/libbeat/metric/system/cgroup/cgv2
../libbeat/metric/system/cgroup/cgv2/io_helper_linux.go:55:16: constant 18446726481523507200 overflows uint32
../libbeat/metric/system/cgroup/cgv2/io_helper_linux.go:56:54: constant 17592184995840 overflows uint32
../libbeat/metric/system/cgroup/cgv2/io_helper_linux.go:57:15: invalid operation: curMajor == major (mismatched types uint32 and uint64)
../libbeat/metric/system/cgroup/cgv2/io_helper_linux.go:57:36: invalid operation: curMinor == minor (mismatched types uint32 and uint64)
```

* Update io_helper_linux.go

Fixing unnecessary type castings

andresrc committed:
Forward port 7.16.1 changelog to master (#29409)

* Forward port 7.16.1 changelog to 8.0 (#29408)

* docs: Prepare Changelog for 7.16.1 (#29404)

* docs: Close changelog for 7.16.1

* Review and cleanup

* Additional cleanup

Co-authored-by: Andres Rodriguez <andres.rodriguez@elastic.co>
(cherry picked from commit 0f81eea724195c72778ac5d8fff5737b7befdd31)

* Review and cleanup

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>
(cherry picked from commit 20aaab20e11b10171382c577606965c0ce584e4b)

* Review and cleanup

ChrsMark committed:
Update k8s library (#29394)

Signed-off-by: chrismark <chrismarkou92@gmail.com>

efd6 committed:
x-pack/filebeat/input/netflow: record IPv6 src and dst addresses (#29383)

legoguy1000 committed:
Move `fips_enabled` setting to AWS Common (#28899)

v1v committed:
[mergify]: forwardport into the master branch (#29390)

ChrsMark committed:
Add option to skip older k8s events (#29396)

michel-laterman committed:
elastic-agent elasticsearch CA fingerprint support (#29128)

* initial commit

* Update Fingerprint method

* Switch to using newly added CATrustedFingerprint attribute

* rename flag

* Fix broken flag

* Add CHANGELOG

* Add container support

michel-laterman committed:
elastic-agent diagnostics pprof (#28798)

* Allow -httpprof to bind to sockets/pipes

* Enable pprof debug endpoint on socket for agent and beats

Force the elastic-agent and all beats that it starts to run the
http/pprof listener on a local socket.

* Add new Pprof command to control.proto

* Add pprof option to diagnostics collect

* Fix linting issues

* Add diagonstics pprof command allow pprof to collect from agent

* Revert debug socket changes

* Cleanup timeout handling

Change pprof timeouts from 2*pprofDur to 30s+pprofDur. Remove timeouts
from the socket requester client as cancellations for long running
requests will be handled by the passed ctx.

* Fix linting issue add timeout flag

Fix linting issues with new command. Add a timeout flag for when pprof
info is gathered. Flag will let users specify the command timeout value.
This value whould be greater then the pprof-duration as it needs to
gather and process pprof data.

* Add more command help text.

* Add CHANGELOG

* move spec collection for routes to fn

* add monitoringCfg reference to control server

* elastic-agent server only processes pprof requests when enabled

* Fix error message fix commands only on elastic-agent

* Add pprof fleet.yml, fix nil reference

* Change pprof setting name to monitoring.pprof.enabled

Chagne the setting in elastic agent from agent.monioring.pprof to
agent.monitoring.pprof.enabled so that policy updates (such as the one
that occurs when the agent is starting in fleet mode) do not use the
default false value if the user has injected the ssetting into fleet.yml

shahzad31 committed:
[Heartbeat] Add run_once mode in the scheduler (#29282)

This PR fixes data stream mode while running heartbeat in run_once mode , before this if you would run heartbeat in run_once mode, it would still push documents to normal heartbeat indexes.

This PR also fixes data like tags not being pushed as part of documents.

kush-elastic committed:
Updates based on comments

legoguy1000 committed:
Add default region config to AWS (#29415)

* Add default regon config to AWS

* update changelog

* Add tests

Commit:
[Automation] Update go release version to 1.17.5 (#29384)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

belimawr committed:
CA trusted fingerprint tests (#29347)

This commit adds tests to `trustRootCA` and `makeVerifyConnection`,
the tests aim to cover the usage of `ssl.ca_tursted_fingerprint`.

Some existing tests were refactored, `openTestCerts` receives a `testing.TB`
and fail tests in case of an error

Commit:
[Automation] Update elastic stack version to 8.1.0-d8a3a806 for testing (#29463)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

mtojek commented on x-pack/filebeat/input/httpjson/config.go:
```diff
@@ -17,6 +17,7 @@ type config struct {
 	Request  *requestConfig  `config:"request" validate:"required"`
 	Response *responseConfig `config:"response"`
 	Cursor   cursorConfig    `config:"cursor"`
+	Chain    []chainsConfig  `config:"chain"`
```
nit: Chain or Chains (struct is named `chainsConfig` hence asking)

kush-elastic replied:
Ans: Chain
It's just, the chain will be an array of steps.

mtojek replied:
Should it be `chainConfig` instead?

mtojek commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,103 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package httpjson
```
It's a general comment, but still - `json_parser.go` contains a lot short named variables: sr, strArr, newIn, bNew.

Could you please reiterate this code and try to make it easier to read? It's really easy to get lost here

kush-elastic replied:
Sure, will update it

belimawr replied:
@kush-elastic have you updated it already?

kush-elastic replied:
Yes, I also added a few examples.

mtojek commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,103 @@
+				return nil, fmt.Errorf("error while parsing json: %v", err)
+			}
+			for _, ir := range newIn.([]interface{}) {
+				if reflect.TypeOf(ir).String() == "string" {
```
Did you borrow this code from some third party library or did you write it?

I'm trying to understand why is this comparison required. You may need to add comments in this function.

kush-elastic replied:
I wrote it and will add comments. Thanks

mtojek commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,103 @@
+)
+
+// getJSON returns array of string values from json string
+func getJSON(b, str string) ([]string, error) {
```
It works similarly to jq, so maybe call it `parse()` or `query()`?

kush-elastic replied:
Got it.

mtojek commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// jsonArr returns array of interface value from interface
+func jsonArr(sr string, bNew interface{}) ([]interface{}, error) {
```
Same story:

Please rename the function and describe the usage in the comment (maybe add a sample?)

kush-elastic replied:
Got it

mtojek commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,103 @@
+
+// jsonNorm returns interface value from interface
+func jsonNorm(sr string, bNew interface{}) (interface{}, error) {
+	if reflect.TypeOf(bNew).String() != "map[string]interface {}" {
```
Are there any other options not to depend on reflection or not really?

kush-elastic replied:
Not in this case. I tried without it but it's causing panic.

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,106 @@ func newRequester(
+func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher) error {
+	var (
+		httpResp       *http.Response
+		ArrayhttpResp  []*http.Response
```
Do we need to expose `ArrayhttpResp` or can it be package-visible only?

kush-elastic replied:
Oh sorry, my mistake.
Thanks

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,106 @@ func newRequester(
+}
+
+func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher) error {
+	var (
+		httpResp       *http.Response
+		ArrayhttpResp  []*http.Response
+		finalHttpResps []*http.Response
+		st             []string
+		err            error
+		b              []byte
+		url            string
```
Please add some comments and rename variables to something more meaningful (st, b, finalHttpResps).

kush-elastic replied:
Got it.

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,106 @@ func newRequester(
+		b              []byte
+		url            string
+	)
+	// perform each requestFactory
```
I guess the next line is self-explanatory and doesn't need this comment :)

kush-elastic replied:
Yeah, Right.

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,106 @@ func newRequester(
+		if i+1 < len(r.requestFactory) {
+			st = nil
+			// collect ids from all responses
+			for _, resp := range ArrayhttpResp {
```
What do you think about testing the code with any mocks returning HTTP errors?

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,106 @@ func newRequester(
+						return err
+					}
+				}
+				// Restore the io.ReadCloser to its original state
+				resp.Body = ioutil.NopCloser(bytes.NewBuffer(b))
```
I admin I got lost here, why is necessary to reset it? Is there an option to simplify this logic?

kush-elastic replied:
Why I did it is because I don't want to store buffered value in the same state.

mtojek commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,106 @@ func newRequester(
+
+}
+
+func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher) error {
```
The function `doRequest` is really long now, what do you think about splitting it into smaller chunks and naming and commenting accordingly?

kush-elastic replied:
Yes i will do that

legoguy1000 committed:
[Filebeat] Enable dynamic inputs (TCP) for Cisco syslog modules (#26159)

- Add tcp option to asa, ftd & ios filesets
- Add SSL option

Closes #28821

Co-authored-by: Lee E. Hinman <lee.e.hinman@elastic.co>

taylor-swanson committed:
[Winlogbeat] Add support for custom XML queries (#29330)

- Added new configuration field (xml_query) to support custom XML queries
- This new configuration item will conflict with existing simple
query configuration items (ignore_older, event_id, level, provider)
- Validator has been updated to check for key conflicts and XML syntax,
but does not check for correctness of XML schema.
- Added unit tests for config validation
- Added unit/system test for XML query runner

rdner committed:
Drop event batch when get HTTP status 413 from ES (#29368)

To prevent infinite loops when having `http.max_content_length` set
too low or `bulk_max_size` too high we now handle this status code
separately and drop the whole event batch producing a detailed error
message on the console.

ChrsMark committed:
Enhance filter check in k8s event metricset (#29470)

v1v committed:
ci: enable orka (#29426)

andrewkroh committed:
Stop calling ChangeMessageVisibility after ReceiptHandleIsInvalid (#29480)

Stop the keepalive goroutine after ErrCodeReceiptHandleIsInvalid is returned by ChangeMessageVisibility.

Add `message_receipt_time` to log messages associated with processing of a given SQS message.

Fix incorrect error being wrapped when ApproximateReceiveCount threshold is reached.

andresrc committed:
Add backport rule for 7.17 branch (#29479)

andresrc committed:
Cleanup breaking changes section (#29490) (#29492)

(cherry picked from commit d9847d75ef2f681905ee0ef8f4913655ec20bada)

klacabane committed:
[monitoring] bump index name version (#29493)

kush-elastic committed:
Update on httpjson and nits

klacabane committed:
[kibana module] write to rootfields instead of modulefields (#29494)

kvch committed:
Change renaming `@timestamp` to copying values in Ingest pipelines (#29425)

jlind23 committed:
Create elastic-agent-project-board.yml (#29500)

jlind23 committed:
Update elastic-agent-project-board.yml

jlind23 committed:
Update elastic-agent-project-board.yml

jlind23 committed:
Update elastic-agent-project-board.yml

dplavcic committed:
[Docs] Update logging.files.permissions documentation to consider umask (#20584) (#28347)

Changes implemented in the (#14119) made all Beats-created files and folders apply an umask of 0027 (on POSIX systems).

Co-authored-by: dplavcic <dplavcic@users.noreply.github.com>

legoguy1000 committed:
Fix cloudtrail config (#29450)

andrewvc committed:
[heartbeat] Only add monitor.status to browser summary events (#29460)

we now only add monitor status to summary events for browsers. Additionally, this renames the poorly renamed wrappers/monitor* to wrappers/wrappers*

v1v committed:
update beats-tester main branch (#29496)

Commit:
[Automation] Update elastic stack version to 8.1.0-befff95a for testing (#29535)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

andresrc committed:
Forward port 7.16.2 changelog to master and cleanup (#29532)

v1v committed:
[mergify]: keep the house clean for conflicts in automated bumps (#29455)

v1v committed:
jjbb: enable 7.17 (and 7.18/19) (#29522)

v1v committed:
update e2e-testing main branch (#29527)

v1v committed:
update ingest-dev main branch (#29525)

MichaelKatsoulis committed:
Calculate memory.working_set.limit.pct for pod and container metricset (#29547)

* Calculate memory.working_set.limit.pct for pod and container metricset

Commit:
add/use blocking wait for test data in statsd (#29543)

* add/use blocking wait for test data in statsd

* refactor to avoid race condition in TestData

* fix serverStarted flag

MichaelKatsoulis committed:
Populate container.id field in container metricset (#29560)

* Populate container.id field in container metricset

dedemorton committed:
Update monitoring-internal-collection.asciidoc (#27624) (#29551)

just to add a clear description regarding cluster_uuid

Co-authored-by: Daisuke Harada <1519063+dharada@users.noreply.github.com>

kvch committed:
Follow up changes in documentation after #28927 (#29501)

* Follow up changes in the docs for logging
* add reference configuration

cmacknz committed:
Save docker-compose logs from integration tests. (#29504)

Save docker-compose logs from system and integration tests.

Allows auditing service logs for issues, in particular it enabled
auditing the elasticsearch logs for deprecation warnings.

The docker log files are now saved under
build/system-tests/docker-logs and preserved as artifacts in
Jenkins.

Co-authored-by: Victor Martinez <VictorMartinezRubio@gmail.com>

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -88,12 +89,15 @@ type requestFactory struct {
+func newRequestFactory(config *requestConfig, configChain []chainsConfig, authConfig *authConfig, log *logp.Logger) []*requestFactory {
 	// config validation already checked for errors here
+	var rf []*requestFactory
 	ts, _ := newBasicTransformsFromConfig(config.Transforms, requestNamespace, log)
-	rf := &requestFactory{
+	// regular call requestFactory object
+	rfs := &requestFactory{
```
The slice can be called `rfs` and the single `requestFactory` can be called `rf`. This will make the code easier to understand.

kush-elastic replied:
Got it. 
Thanks

belimawr commented on x-pack/filebeat/input/httpjson/chain.go:
```diff
@@ -0,0 +1,16 @@
+	Request  requestConfig  `config:"request"`
+	Response responseConfig `config:"response,omitempty"`
+	Replace  string         `config:"replace,omitempty"`
+}
```
Those structs are a single thing, their name can be in the singular too:
```suggestion
package httpjson

type chainConfig struct {
	Step stepConfig `config:"step" validate:"required"`
}

type stepConfig struct {
	Auth     authConfig     `config:"auth,omitempty"`
	Request  requestConfig  `config:"request"`
	Response responseConfig `config:"response,omitempty"`
	Replace  string         `config:"replace,omitempty"`
}

```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -88,12 +89,15 @@ type requestFactory struct {
 }
 
-func newRequestFactory(config *requestConfig, authConfig *authConfig, log *logp.Logger) *requestFactory {
+func newRequestFactory(config *requestConfig, configChain []chainsConfig, authConfig *authConfig, log *logp.Logger) []*requestFactory {
```
Because the new argument is a slice, its name should be plural
```suggestion
func newRequestFactory(config *requestConfig, configChains []chainsConfig, authConfig *authConfig, log *logp.Logger) []*requestFactory {
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,13 +164,13 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
 func newRequester(
 	client *httpClient,
-	requestFactory *requestFactory,
+	requestFactory []*requestFactory,
```
Same thing here regarding plurals: `requestFactories`

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
 
-	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, httpResp)
+// generateNewUrl returns new url value using replacement from oldUrl with str ids
+func generateNewUrl(replacement string, oldUrl string, str []byte) (url.URL, error) {
```
If `str` contains IDs (and because it's a `[]byte`), it could be renamed to something like `ids`

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+	)
+	for i, rf := range r.requestFactory {
+		// iterate over collected ids from last response
+		if i != 0 {
```
What about inverting the logic here so the first block of code is also the first to be executed? I believe it will improve readability.
```suggestion
		if i == 0 {
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+			// perform and store regular call responses
+			httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+			if err != nil {
+				return fmt.Errorf("failed to execute http client.Do: %w", err)
```
There seems to be a mismatch between the comment and the function called. You could either put the name of the function that is actually called, or just a more generic message stating there was a failure while executing the request.
```suggestion
				return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+			if err != nil {
+				return fmt.Errorf("failed to execute http client.Do: %w", err)
+			}
+			if len(r.requestFactory) <= 1 {
```
`len(r.requestFactory)` is never `0` here, so this can be simplified:
```suggestion
			if len(r.requestFactory) == 1 {
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+			urlCopy = rf.url
+			urlString = rf.url.String()
+			// perform request over collected ids
+			for _, s := range ids {
```
You could call the single ID `id`:
```suggestion
			for _, id := range ids {
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+				// collect data from new urls
+				httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+				if err != nil {
+					return fmt.Errorf("failed to execute http client.Do: %w", err)
```
Same thing here as in another comment, the message can better reflect the method called:
```suggestion
					return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+					return fmt.Errorf("failed to execute http client.Do: %w", err)
+				}
+				// store data according to response type
+				if i+1 == len(r.requestFactory) && len(ids) != 0 {
```
Here you could keep the `len` on the same side of both comparisons, I believe it will improve readability
```suggestion
				if len(r.requestFactory)  == i+1 && len(ids) != 0 {
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+			}
+			rf.url = urlCopy
+
+			if i+1 == len(r.requestFactory) {
```
Same thing here, let's be consistent with the other parts of the code:
```suggestion
			if len(r.requestFactory) == i+1 {
```

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+	// collect ids from all responses
+	for _, resp := range intermediateResps {
+		if resp.Body != nil {
+			b, err = ioutil.ReadAll(resp.Body)
```
The `ioutil` package has been deprecated in Go 1.16, you can use `io.ReadAll` instead.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+			}
+		}
+		// Restore the io.ReadCloser to its original state
+		resp.Body = ioutil.NopCloser(bytes.NewBuffer(b))
```
This line and the fact that making sure the response.Body is always closed makes me wonder if `collectResponse` should read the response.Body and return a `[]byte`. I see two advantages on this approach:
1. No doubt about where the response.Body needs to be closed
2. We read and close the response.Body, so the remote serve does not need to wait for when we decide to read it.

I might be missing something, but I also did not see usage of any other fields of the response.

kush-elastic replied:
yeah, I think that's a good idea. what we can do is store []byte for intermediate responses and for final responses as it is. but  I think it will cause problems for pagination. as pagination is using responses to iterate.
I will see maybe there is something I can do.
storing this many responses is not a good idea 
Thanks 👍

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+func (r *requester) processAndPublishEvents(publish int, stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher, finalResps []*http.Response) (int, error) {
+	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, finalResps)
+	if err != nil {
+		r.log.Errorf("error recieving eventCh: %v", err)
```
If there was an error receiving `eventCh`, it does not make sense to try using it a few lines below. Why not return from here?

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -189,15 +346,17 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
-		if err := publisher.Publish(event, trCtx.cursorMap()); err != nil {
-			r.log.Errorf("error publishing event: %v", err)
-			continue
+		if publish == 0 {
```
Could this `publish` parameter be a `bool`? So far I only saw it assuming the values `0` and `1`.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -163,21 +181,160 @@ func newRequester(
+}
+
+// processAndPublishEvents process and publish events based on response type
+func (r *requester) processAndPublishEvents(publish int, stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher, finalResps []*http.Response) (int, error) {
```
By convention `context.Context` should be the first parameter, I suggest to refactor this method signature.

belimawr commented on x-pack/filebeat/input/httpjson/response.go:
```diff
@@ -74,63 +74,68 @@ func newResponseProcessor(config *responseConfig, pagination *pagination, log *l
 }
 
-func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *transformContext, resp *http.Response) (<-chan maybeMsg, error) {
+func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *transformContext, resp []*http.Response) (<-chan maybeMsg, error) {
```
`resp` is a slice, it could be called `resps`
```suggestion
func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *transformContext, resps []*http.Response) (<-chan maybeMsg, error) {
```

belimawr commented on x-pack/filebeat/input/httpjson/response.go:
```diff
@@ -74,63 +74,68 @@ func newResponseProcessor(config *responseConfig, pagination *pagination, log *l
+				trCtx.updateLastResponse(*page)
 
-			rp.log.Debugf("last received page: %#v", trCtx.lastResponse)
+				rp.log.Debugf("last received page: %#v", trCtx.lastResponse)
```
[Question]

Should we add some information about the request in the chain?

kush-elastic replied:
I am not sure but what do you think, should we add like chain URL(generated using ids)?

belimawr replied:
Maybe, or the index from the for loop. The main idea is that when looking at the logs, we know from which request in the chain this log entry was about.

belimawr commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,143 @@
+
+package httpjson
+
+import (
```
In general this file contains a very low level JSON manipulation, I'm just wondering why it is needed.

kush-elastic replied:
I tried to use the Mapstr package but it is not able to parse []interface{} types of data so I wrote this code to collect those kinds of info from JSON.

belimawr replied:
Unfortunately there is no way to parse any valid JSON as a `map[string]interface{}`. One thing you can do is to unmarshal the JSON into a empty interface (`interface{}`) and then do a type assertion/type switch to manipulate it. [I made a quick example in the Go Playground](https://go.dev/play/p/rmqk9-KHDqS) where you can see different JSONs being parsed into a empty interface and assuming different types.

I would like to better understand the need for this special JSON manipulation because the current implementation mixes JSON parsing with some trasversing/filtering. Look at the [`parse`](https://github.com/kush-elastic/beats/blob/664d95b1b043b3199e88156f72c488928bb175f3/x-pack/filebeat/input/httpjson/json_parser.go#L15) function: it receives a JSON and a key, splits the key on the `.` (so it assumes a specific JSON structure), does some data manipulation and returns a `[]string`. That is definitely not a generic JSON parser.

This makes me wonder if this whole JSON manipulation belongs to the `httpjson` package or should be implemented somewhere else.

There are also some hard to grasp constructions and many type casts that could lead to bugs/panics.

Would you like to setup a call to discuss this PR face-2-face?

kush-elastic replied:
Thanks for the detailed analysis.
Let's connect at your convenience, that would be helpful.

belimawr reviewed (changes requested):
I added quite some comments regarding readability and naming things, let me know what you think.

kush-elastic committed:
Resolve comments from Tiago

ChrsMark committed:
Add support for latest k8s versions (#29575)

klacabane committed:
elasticsearch module - xpack enabled parity (#29548)

* elasticsearch module - xpack enabled parity

* flip ActiveOnly default

* add ml_job's node

* lint

* update shard cluster's state path

* add back state.id to prevent breaking changes

tetianakravchenko committed:
Align elastic-agent-standalone manifest with the kubernetes package changes (#29595)

* align elastic-agent-standalone manifest with the managed version

Signed-off-by: Tetiana Kravchenko <tetiana.kravchenko@elastic.co>

* revetn docker image version

Signed-off-by: Tetiana Kravchenko <tetiana.kravchenko@elastic.co>

* remove ES_HOST used to run test locally

Signed-off-by: Tetiana Kravchenko <tetiana.kravchenko@elastic.co>

* set default values for container parser implicitly

Signed-off-by: Tetiana Kravchenko <tetiana.kravchenko@elastic.co>

* remove skip_older as it is a default value anyway

Signed-off-by: Tetiana Kravchenko <tetiana.kravchenko@elastic.co>

michel-laterman committed:
Remove references to username/password (#29458)

* Remove references to username/password

* restore ouput username/password

* Update CHANGELOG

legoguy1000 committed:
[Filebeat] httpjson - Add function to return User-Agent #29453 (#29528)

Adds a new value template function for the httpjson input to generate/update the User Agent.

klacabane committed:
beat/kibana module - add {module}.elasticsearch.cluster.id property (#29577)

* add elasticsearch.cluster.id property to beats state

* add elasticsearch.cluster.id property to beats stats

* fix test

* lint

* add missing reloads property

* dont break earlier assumptions

* remove trailing comma

* update beats.stats fields

* add updated fields doc

* beat: update cluster.id path

* kibana: update cluster.id path

* mage fmt update

* fix fields path

* add missing map initialization

* add changelog entry

* fix pr number

Commit:
[Automation] Update elastic stack version to 8.1.0-b989cb5c for testing (#29618)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

kwinstonix committed:
Fix gcp metrics metricset apply aligner to all metric_types (#29513) (#29514)

Co-authored-by: endorama <526307+endorama@users.noreply.github.com>

v1v committed:
ci: enable 7.17 branch (#29602)

klacabane committed:
kibana.stats: extract correct index property (#29622)

* kibana.stats: extract correct index property

* changelog entry

marclop committed:
apm-server: Remove `data_streams.enabled` flag (#29168)

* apm-server: Remove `data_streams.enabled` flag

Removes the `"apm-server.data_streams.enabled` flag in the `apm-server`
spec, since it's been marked as deprecated in `8.0`.

Signed-off-by: Marc Lopez Rubio <marc5.12@outlook.com>

Commit:
[Automation] Update elastic stack version to 8.1.0-c1a942c7 for testing (#29630)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

klacabane committed:
add elasticsearch.cluster.id to logstash metricsets (#29625)

* add elasticsearch.cluster.id to logstash metricsets

* changelog entry

adriansr committed:
Ensure that match_only_text and wildcard fields can be added to default_field list (#29634)

Fixes the construction of the index template's default_field list so that fields of type match_only_text
and wildcard can be added to it.

belimawr commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,143 @@
+)
+
+// parse returns array of string values from json string
+func parse(jsonbyte, keyField string) ([]string, error) {
```
@kush-elastic thanks a lot for answering my questions and clarifying about the feature!

It definitely makes sense to have this feature here, but we can make the code easier to read and better leverage the Go type system to provide more safety.

Some things we can improve:
 - avoid adding the type to variable names, specially if they don't match. Ex: avoid things like `jsonbyte, keyField string`
 - `keyField` can just be `key`
 - give variable clear names, avoid things like `inf` and `ir`
 - don't do type casting like `myVar.(string)` because this can panic, choose the `castVar, ok := myVar.(string)` instead. If the type cast is possible, `ok` will be set to `true`
 - you can umarshal the JSON data into a empty interface variable (`var data := interface{}`) and then do type assertions to know if you're dealing with JSON objects or arrays
 - When returning errors with some extra message, use the formatting directive `%w` to wrap them, like this: `fmt.Errorf("Ooops, there was an error: %w", err)`

I wrote a [draft of your parse function](https://go.dev/play/p/_tLizvZvnqT) on Go playground applying some of the ideas I mentioned above. What do you think?

kush-elastic replied:
Thanks, it really helped a lot. 👍

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +164,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	}
+
+	httpResp.Body.Close()
+	r.log.Infof("request finished: %d events published", n)
```
I believe this should be on level debug

kush-elastic replied:
it's from the original code. I am not sure why it is showing in diff.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +164,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	if err != nil {
+		return url.URL{}, fmt.Errorf("failed to create regex on provided value: %w", err)
+	}
+	newUrl, err := url.Parse(string(reg.ReplaceAll([]byte(oldUrl), ids)))
```
You can use `ReplaceAllString` instead
```suggestion
	newUrl, err := url.Parse(reg.ReplaceAllString(oldUrl, ids))
```

Maybe it also makes sense to have `ids` as a string? So this function has got all types matching and the caller is the responsible for doing type casts, if needed.

kush-elastic replied:
Got it.
Thanks

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +164,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, finalResps)
 	if err != nil {
-		return err
+		return 0, fmt.Errorf("error recieving eventCh: %v", err)
```
You should use `%w` here, also I believe we can make this error message more clear by describing what went wrong instead of saying what variable is not correctly populated.
```suggestion
		return 0, fmt.Errorf("error starting response processor: %w", err)
```

kush-elastic replied:
Sure, I will update it.
Thank for comment

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +164,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+				if err != nil {
+					return err
+				}
+				continue
```
Because `rfSize == 1` that's the last iteration of the for loop, so you can return early here. Because this method is already quite long, this helps a lot to understand the flow of the program.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +164,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+			}
+			rf.url = urlCopy
+
+			if rfSize == i+1 {
```
You want this to be true in the last of the slice, right?

What about flipping the condition to make it more explicitly we're looking for the last element?
```suggestion
			if i == rfSize-1 {
```

What do you think? Does it help readability?

kush-elastic replied:
Yes
Right
That's a good idea
It makes sense
Thanks

efd6 committed:
x-pack/auditbeat/module/system/socket: get full length path and arg from /proc when not available from kprobe (#29410)

Also use first arg from sysinfo.Processes in place of Name to avoid process name
truncation.

efd6 committed:
x-pack/filebeat/module/sophos/xg: fix kv field separation and add support for timestamped log line (#29331)

This change also needed changes in the cisco asa module's test expect files
to bring timestamp handling during testing up to date.

v1v committed:
ci: elastic-agent goIntegTest is not available (#29566)

wilsonge committed:
Add kubernetes job name as the controller (#28954)

kvch committed:
Add more filtering options to journald input (#29294)

## What does this PR do?

This PR adds support for `unit`, `transports` and `syslog_identifiers` options for filtering.

This PR also introduces a breaking change to `include_matches` option. From now on it does not accept a list of expressions. Now both conjunction (AND) and disjunctions (OR) are supported when applying matches to journals.

Collecting entries of two different units:

```yaml
- type: journald
  include_matches.or:
  - equals:
    - _SYSTEMD_UNIT=my_unit
    - _SYSTEMD_UNIT=my_other_unit
```

Collecting entries using syslog transport for a unit
```yaml
- type: journald
  include_matches.and:
  - equals:
    - _SYSTEMD_UNIT=my_unit
    - _TRANSPORT=syslog
```

Although the configuration lets you write complex expressions, systemd does not provide full logical expression support.

## Why is it important?

When this change merged, journald input can be marked either beta or GA. Furthermore, now it provides similar filtering capabilities as the good old community Journalbeat did.

kvch committed:
Add warning about removing Functionbeat (#29295)

belimawr committed:
Update Beats docs to include ES CA fingerprint (#29632)

Update Beats ES connection documentation to include ES CA fingerprint
information.

leehinman committed:
[Winlogbeat] Switch to ingest node processing (#29435)

What does this PR do?

Switches to ingest node processing instead of local javascript
processing for modules.

Why is it important?

For Agent we are using ingest node processing. This keeps the
Winlogbeat and Agent processing in sync and allows bug fixes to be
shared between them.

Changes

- remove JS processing
- remove testing of JS processing
- add ingest node pipelines from agent
- use routing pipeline to direct to correct module pipeline
- embed pipelines in winlogbeat executable
- add `export pipeline` command
- add loading of pipelines to setup command
- add loading of pipelines to every ES connection if the pipelines
  don't exist
- exports several functions from filebeat/fileset to avoid code
  duplication
- update docs
- update default config
- update reference config

Closes #29184

v1v committed:
[packaging] Bump golang-crossbuild version debian8 (#23872)

dedemorton committed:
Fix example command to avoid showing system module under windows (#29569)

Commit:
[Automation] Update elastic stack version to 8.1.0-1f83f9c5 for testing (#29702)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

kush-elastic committed:
address comments and improved json_parser

michel-laterman committed:
Allow elastic agent in containers to use basic auth to get service token (#29651)

* Allow elastic agent in containers to use basic auth to get service token

Allow the agent to use basic auth defined by env vars to retrieve a
service token from Elasticsearch and inject it into the config used for
the agent and fleet.

* Fix linter

* Use Kibana API for all requests.

Change from using the ES api to gather the token to the Kibana API.

dosmanak committed:
docs: reorder kubernetes provider section (#29443)

* docs: reorder kubernetes provider configuration setting section

* docs: reorder kubernetes provider template configuration intro

* docs: fix beatname_lc reference expansion

https://asciidoc-py.github.io/userguide.html#X81

Co-authored-by: DeDe Morton <dede.morton@elastic.co>

* docs: reorder docker provider configuration setting section

* docs: reorder jolokia provider configuration setting section

reword interfaces items description

* docs: reorder nomad provider configuration setting section

* docs: reorder nomad provider template configuration intro

Co-authored-by: DeDe Morton <dede.morton@elastic.co>

Commit:
[Automation] Update elastic stack version to 8.1.0-af114d16 for testing (#29721)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

klacabane committed:
cluster_stats: use correct license value (#29711)

* add integration test

* target correct license property

* fix license path

* cluster_stats: use correct license value

* changelog entry

andrewvc committed:
[Heartbeat] Separate http req per task (#29697)

This is an attempt to fix the race reported in #29580 by instantiating a separate http request per HTTP task. The theory being that the HTTP library modifies the headers and that the req object is not safe to share.

This has passed manual testing using mode: all against endpoints with multiple A records.

Tests are not included here due to the tricky nature of testing here, but we will do so in a follow-up

andrewvc committed:
[Heartbeat] Use timer.reset now that golang has been updated (#29729)

Undoes https://github.com/elastic/beats/pull/27006/files (while preserving the new test), and also cleaning up the syntax using `time.Until`.

Since the The golang bug I reported in https://github.com/golang/go/issues/47329 has been fixed since somewhere in go 1.16.x (it's hard to track the exact version). This should be backported to 7.16.x since that already uses go 1.17.x and is safe.

Commit:
[Automation] Update elastic stack version to 8.1.0-5e490954 for testing (#29737)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>

kvch committed:
Do not mention removal if version is not specified in `cfgwarn` messages (#29727)

adriansr committed:
System/socket: Support kernel_clone() replacement for _do_fork() (#29744)

Updates the system/socket dataset to support kernels 5.10+ where the _do_fork kernel function
is replaced by kernel_clone. This was preventing Auditbeat to start.

adriansr committed:
auditd: Store program arguments in process.args array (#29601)

Changes Filebeat's auditd module to store program arguments
(from an EXECVE call) in process.args (arg0 also in process.executable).
Previously it was using fields arg0 to argN under auditd.log.

This prevents too many fields being created. When a call contained
more than 10.000 arguments, this lead to an ingest error and contributed to
very large indices:

> Could not index event to Elasticsearch: "status"=>400,
>   "error"=>{
>     "type"=>"illegal_argument_exception",
>     "reason"=>"Limit of total fields [10000] has been exceeded"}}

andresrc committed:
Forward port 7.16.3 changelog to master (#29777)

adriansr committed:
Filebeat tests: Restore `@timestamp` field validation (#29772)

This restores the `@timestamp` field in Filebeat's module tests, so that
it is properly validated.

Temporarily disable timestamp validation in ibmmq module

v1v committed:
ci: uses aliases for the branches (#29706)

kvch committed:
Do not add date to index if `@meta.index` is set (#29775)

adriansr committed:
ibmmq: Fix timestamp parsing (#29773)

This fixes the timestamp parsing in ibmmq logs:
- Date processor format definition was broken for ES 8.0 (extra `a`
  character).
- The header date format in some logs was unsupported.
- The Time() field, with correct TZ and higher precission was ignored.

ChrsMark committed:
Add k8s metadata in state_cronjob metricset (#29572)

v1v committed:
jjbb: remove obsoleted branches (<7.16) (#29707)

tetianakravchenko committed:
Remove overriding of index pattern on the Kubernetes overview dashboard (#29676)

* Remove overriding of index pattern on the Kubernetes overview dashboard; align visualisations with changes in https://github.com/elastic/integrations/pull/2151

Signed-off-by: Tetiana Kravchenko <tetiana.kravchenko@elastic.co>

* add PR number in changelog

Signed-off-by: Tetiana Kravchenko <tetiana.kravchenko@elastic.co>

MichaelKatsoulis committed:
Add a readme for k8s autodiscover provider (#28213)

* Add a readme for k8s autodiscover provider

adriansr committed:
Missing changelog entry for #29773 (#29791)

Relates #29773

Commit:
[Automation] Update elastic stack version to 8.1.0-7004acda for testing (#29783)

Co-authored-by: apmmachine <infra-root-apmmachine@elastic.co>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

efd6 committed:
libbeat/processors/add_process_metadata: implement a process cache eviction policy (#29717)

The eviction policy implemented here guarantees a hard limit to the number of entries
in the cache and provides a downward force on the number of entries over time with an
exponential decay on the fraction of entries that are older than the cache expiration
duration.

blakerouse committed:
Change docker image from CentOS 7 to Ubuntu 20.04 (#29681)

* Switch to Ubuntu 20.04.

* Fix Dockerfile.

* Change to amd64 in Dockerfile.tmpl.

* Add missing curl and ca-certificate deps. Change back to x86_64.

* Fix issues with setcap and xz extraction.

* Add changelog, fix remaining issues.

* Fix synthetics deps

* Fix apt-get. Remove todo from packages.yml.

Co-authored-by: Andrew Cholakian <andrewvc@elastic.co>

mergify[bot] committed:
Prepare 8.0.0-rc1 changelog (#29795) (#29806)

(cherry picked from commit 9a5ee416b097a4dbae08bcef7876f37579bcd45c)

Co-authored-by: Andres Rodriguez <andres.rodriguez@elastic.co>

lucasfcosta committed:
Add summary to journeys which don't emit journey:end (early node subprocess exits) (#29606)

* update link to beats developer guide

* fix: add summary to journeys which don't emit journey:end [fixes #28770]

* fix: avoid cmd/status when journey has already finished

kush-elastic committed:
Add entry to CHANGELOG.next.asciidoc

kush-elastic committed:
Merge branch 'master' into salesforce_httpjson_chain_calls

kush-elastic committed:
mage update and mage check for ci linting

kush-elastic commented:
There is a new PR to elastic:master branch.
please refer to it and ignore this one.
[here](https://github.com/elastic/beats/pull/29816) is the link.
